[FIX] stock: print the delivery address on slip when using packages

mathcoutant opened the pull request:
### Issue
The delivery slip of a picking with a package doesn't print the delivery address of the partner.

### Steps to reproduce:
1. Go to Settings > Inventory > Operations and enable 'Packages'
2. Go to Inventory > Configuration > Warehouse Management > Operations Types
3. Open operation type 'Delivery Orders' and enable 'Show Detailed Operations' and 'Move Entire Packages'
4. Create a package: in Purchase, new RFQ, select products, confirm order, receive products, click on 'Put in Pack' and validate
5. In Inventory create a new Delivery Order, select a Delivery Address and the package you just created
6. Mark it as Done, validate and print the delivery slip
7. The delivery address is not the right one (the one selected on step 5)

### Cause:
When creating the moves for the items in the package, the partner_id is not given. So the moves don't have the field partner_id. When the method should_print_delivery_address() of stock_picking is called, it returns False as self.move_ids[0].partner_id is undefined, and the delivery address is not printed.
This bug does not occur if action_confirm is called before setting the partner because when setting the partner the write method will change the partner_id attribute of the moves.

### Solution:
Add the partner_id in _generate_moves() of stock_package_level to create the moves with the selected partner_id.

opw-3901104

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/165194).

adwid reviewed (changes requested):
Can you add a test ? 
I will then look at the fix :) 

You can simplify the step 4 by updating the inventory with the data you want (that way you avoid the process of a PO + receipt)
I let you check the use of the method `_update_available_quantity` in the tests

ticodoo commented on addons/stock/models/stock_package_level.py:
```diff
@@ -145,6 +145,7 @@ def _generate_moves(self):
                         'location_dest_id': package_level.location_dest_id.id,
                         'package_level_id': package_level.id,
                         'company_id': package_level.company_id.id,
+                        'partner_id': self.picking_id.partner_id.id,
```
```suggestion
                        'partner_id': package_level.picking_id.partner_id.id,
```

ticodoo reviewed (commented):
I'm a bit confused, by a few things:
- does this bug not occur happen in v15? (I think the logic is the same, but I didn't check)
- I thought we already fixed this bug with https://github.com/odoo/odoo/pull/107814 ? (did it get reintroduced somehow?)
- won't the bug still happen if the partner is set after the `action_confirm` is called? (since `_generate_moves()` is called by it) If yes, then your fix will probably especially not work in saas-16.3/ since `action_assign` was refactored to call `_generate_moves()` 

If the answer to all 3 things is _nope_, then the fix is ok for me with my code change recommendation so that it doesn't break for multi-records :+1:

mathcoutant commented:
> I'm a bit confused, by a few things:
> 
> * does this bug not occur happen in v15? (I think the logic is the same, but I didn't check)
> * I thought we already fixed this bug with [[FIX] stock: delivery address for entire packages #107814](https://github.com/odoo/odoo/pull/107814) ? (did it get reintroduced somehow?)
> * won't the bug still happen if the partner is set after the `action_confirm` is called? (since `_generate_moves()` is called by it) If yes, then your fix will probably especially not work in saas-16.3/ since `action_assign` was refactored to call `_generate_moves()`
> 
> If the answer to all 3 things is _nope_, then the fix is ok for me with my code change recommendation so that it doesn't break for multi-records 👍

For the first point, yes this bug also happen in v15 (my bad, I did not check), I changed the targeted version.

For the two others, I guess I was not clear in the PR comment. 
The fix you are referring to prints the delivery address for packages whose moves have a partner_id set to something, which was not the case before as it was only checking moves without packages. 

This PR fixes another bug:
- When calling action_confirm and then setting the partner, everything works fine because the moves are generated by action_confirm and then the write method is called and sets the partner_id on the moves.
- So the bug this PR is trying to fix only occurs when action_confirm is called with the partner set, in this case the moves are created without partner_id and the write method is not called. So in the end, the moves do not have a partner_id and the delivery address is not printed.

Also, this bug is not present in saas-16.3 and further versions.

Thanks for the change recommendation 👍

ticodoo commented:
@mathcoutant thanks for the detailed explanation!

@robodoo r+
